### PR TITLE
Add RAG functionality to Copilot

### PR DESCRIPTION
**Pull request by K-Schubert**
**Overview**

Added a simple RAG functionality to the EAK-Copilot.

**Issue Reference:**

- #79 
- #105 
- #107 
- #97 
- #134 
- #110 
- #135 
- #112 

**Changes Made**

- Added rag/app/main.py to implement the /init_expert, /get_docs, /rerank and /embed endpoints.
- Added rag/app/models.py to implement the ResponseBody and RAGRequest model classes.
- Added rag/app/test_semantic_search.py to run a simple semantic search query.
- Added rag/Dockerfile to build the rag image.
- Added rag/requirements.txt.
- Updated docker-compose.yml to include the rag service.
- Updated .env.example to include the postgresql port number.
- Updated the README.md.

**Testing**

- Tested endpoints manually through swagger docs at localhost:8010/docs.
- Removed the db/data folder to index docs from scratch.
- Ran rag/app/test_semantic_search.py.

**Screenshots**

None

**Additional Notes**

- Crawling/Scraping/Indexing of fedlex.ch, ahv-iv.ch, zas.admin.ch and bsv.admin.ch will be implemented in a future sprint. For the moment only dummy data is indexed.
- Auto data crawling/scraping/indexing will be configured from the main config.yaml file in the future (for the moment this process is manual).

**Requested Reviewers**

@tabee

### CODE REVIEW

File: api/requirements.txt
Comment (tabee) [inline api/requirements.txt:1]: Do we have a process where we check whether the current copilot can also use the latest requirements? It is certainly good to specify version numbers in the main branch. But we also need to update regularly

File: docker-compose.yml [hunk 1]
```diff
@@ -62,7 +61,7 @@ services:
   # frontend
   chatgpt-minimal:
     build:
-      context: https://github.com/CdC-SI/chatgpt-minimal.git#main
```
Comment (tabee) [inline docker-compose.yml:65]: I hope soon we onlay can clone from #main

File: rag/app/main.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    Dummy endpoint for text embedding.
+    """
+    return {"message": "Embed data test!"}
+
+@app.get("/rerank", summary="Reranking endpoint", response_description="Welcome Message")
+async def rerank():
+    """
+    Dummy endpoint for retrieved docs reranking.
+    """
+    return {"message": "Rerank test!"}
```
Comment (tabee) [inline rag/app/main.py:132]: alternative:
``` python 
return Response(content="Not Implemented", status_code=status.HTTP_501_NOT_IMPLEMENTED)
```
oder
``` python 
{"message": "Rerank Not Implemented"}
```

File: rag/app/test_semantic_search.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+# Database connection parameters
+DB_PARAMS = {
+    "user": POSTGRES_USER,
+    "password": POSTGRES_PASSWORD,
+    "database": POSTGRES_DB,
+    "host": POSTGRES_HOST,
+    "port": POSTGRES_PORT,
+}
+
```
Comment (tabee) [inline rag/app/test_semantic_search.py:23]: do we have to store repetitive code in an external file?

**Review by tabee — APPROVED** [review]

File: api/requirements.txt
Comment (K-Schubert) [reply api/requirements.txt:1]: Not yet. I think it's safest to hardcode versions in the requirements.txt, and either use tools such as renovate bot (checks whether dependencies are outdated, compatible with github and bitbucket) or a check in the CI/CD pipeline with a "pip list --outdated" to notify developers. Implementing the CI/CD pipeline will be for next sprint I guess.

File: rag/app/test_semantic_search.py (hunk 3 above)
Comment (K-Schubert) [reply rag/app/test_semantic_search.py:23]: Refactored this with /config.py at root which stores all env variables. Had to refactor the docker-compose.yml as well as individual service dockerfiles. Implemented in 44697b00916c95a35fc0c14b7d669a8005af9d73 and 065d019a38589e924a0c507f545fe5e1461aa94a.

**Pull request merged**